### DISCUSSION TITLE
Parsing out data from Nornir output

### DISCUSSION THREAD

**Discussion by mikeymastro** [nornir-automation/nornir#714] [General]
**edit - this is for Nornir 3.0**

Hello all.

I want to apologize up front if this is a super simple task, as i'm extremely new to Nornir and trying to learn as much as possible.

I am running netmiko_send_command of _**show version**_ against all of my Arista devices, and I am just looking to parse out the version number from each device.  I cannot slice it like I normally would within Python, most likely due to the data structure that is returned from Nornir.  I am having a really difficult time trying to figure this out.  I tried to search online for quite some time but to no avail.

I also tried using the Napalm library using the _**get_facts**_ but still no luck.  I think if I can figure out one, I could use it with the other (i'd think!).

Any help would be appreciated.  And as it pertains to network automation as a whole, i'm an extreme amateur so again, I apologize if I don't quite understand some of the responses.  I hope to get to an intermediate state as quickly as possible haha. 

Thanks all.

**Comment by itheodoridis** [comment]
Hello. 

Are you having an issue accessing the object returned by nornir?

It’s usually a multiresult object. You have to use some references in it in order to get to the data.

Is this your problem?

 

 

-----------------------------------------------------

        Ioannis Theodoridis

   Network & Systems Engineer

            ---------

         Networks Section

         Informatics Division

           Bank of Greece

-----------------------------------------------------

 

From: magic9669 ***@***.***> 
Sent: Friday, August 27, 2021 3:58 PM
To: nornir-automation/nornir ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [nornir-automation/nornir] Parsing out data from Nornir output (#714)

 

Hello all.

I want to apologize up front if this is a super simple task, as i'm extremely new to Nornir and trying to learn as much as possible.

I am running netmiko_send_command of show version against all of my Arista devices, and I am just looking to parse out the version number from each device. I cannot slice it like I normally would within Python, most likely due to the data structure that is returned from Nornir. I am having a really difficult time trying to figure this out. I tried to search online for quite some time but to no avail.

I also tried using the Napalm library using the get_facts but still no luck. I think if I can figure out one, I could use it with the other (i'd think!).

Any help would be appreciated. And as it pertains to network automation as a whole, i'm an extreme amateur so again, I apologize if I don't quite understand some of the responses. I hope to get to an intermediate state as quickly as possible haha.

Thanks all.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub <https://github.com/nornir-automation/nornir/discussions/714> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADVIRDOS7SQYZKTPMX6GU3LT66DULANCNFSM5C5P36XQ> .
Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>  or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub> .   <https://github.com/notifications/beacon/ADVIRDJAQ7APZPF5NKE2DJLT66DULA5CNFSM5C5P36X2YY3PNVWWK3TUL52HS4DFVJCGS43DOVZXG2LPN2VGG33NNVSW45C7NFSM4ABWBZKA.gif>

**Reply by mikeymastro** [reply]
I believe so, if I understand your question correctly.  With Python for example, I can index a dictionary key and get the value returned.  I am looking to do the same for the software version.  Below is some of the code to use an as example

```
Router = nr.filter(F(groups__contains="arista"))
result = Router.run(netmiko_send_command, command_string="show version")
print_result(result)
```

The output returned has this line within it:  **_Software image version: 4.21.6F_**

I just want to parse out that _**4.21.6F**_ portion.  Unfortunately, with the print_result function, I can't just add _**['Software image version:']**_ like I am used to doing within Python itself.  I am most likely just missing something rather easy, but I have tried everything and cannot get it at all for some reason.

**Reply by dbarrosop** [reply]
Hi! The solution to your problem is in the docs: https://nornir.readthedocs.io/en/latest/tutorial/task_results.html

**Reply by mikeymastro** [reply]
I will check that out.  Thank you very much.  I appreciate it.

And I apologize for the slow responses.  I am out of town with minimal access to the internet currently.  I say this because I want to try and implement this as quickly as possible and I can update this thread.  Thanks again.

**Reply by mikeymastro** [reply]
I am not so sure that link provides the solution I am looking for, but then again, it may very well be my inexperience with this tool as well.  I tried several things within that doc and while i'm able to index the result itself, I cannot seem to pull just the pertinent information I need to pull.  

Here is an example output from one device of mine:

```
vvvv r1-sandbox: netmiko_send_command ** changed : False vvvvvvvvvvvvvvvvvv INFO
Arista DCS-7280SR-48C6-F
Hardware version:    11.03
Serial number:       SSJ17346121
System MAC address:  2899.3aaf.210d

Software image version: 4.21.6F
Architecture:           i386
Internal build version: 4.21.6F-12355585.4216F

Uptime:                 74 weeks, 3 days, 9 hours and 29 minutes
Total memory:           7825900 kB
Free memory:            6110428 kB

^^^^ END netmiko_send_command ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I used some of the options within the doc link listed above, and printing out the type of the output is coming back as _**<class 'nornir.core.task.Result'>**_

I don't know if you can convert this to another object type or what not, but I cannot figure out how to index or parse out the _**Software image version:**_ so all that is returned is the _**4.21.6F**_

I hope this makes sense for everyone.  I may be using the wrong terminology and such, so my apologies if this is the case.  Thank you all for the assistance thus far.  I really appreciate it

**Reply by ktbyers** [reply]
@magic9669 What are you trying to do currently? I.E. what is your code that you are using and what issue are you running into?
